solr/core/src/java/org/apache/solr/core/XmlConfigFile.java

```diff
@@ -125,59 +75,30 @@ public XmlConfigFile(
    */
   public XmlConfigFile(
       SolrResourceLoader loader,
-      Function<String, InputStream> fileSupplier,
       String name,
       InputSource is,
       String prefix,
       Properties substituteProps)
       throws IOException {
     if (null == loader) throw new NullPointerException("loader");
     this.loader = loader;
-
     this.substituteProperties = substituteProps;
     this.name = name;
     this.prefix = (prefix != null && !prefix.endsWith("/")) ? prefix + '/' : prefix;
-    try {
-      javax.xml.parsers.DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
-
-      if (is == null) {
-        InputStream in = fileSupplier.apply(name);
-        if (in instanceof ZkSolrResourceLoader.ZkByteArrayInputStream) {
-          zkVersion = ((ZkSolrResourceLoader.ZkByteArrayInputStream) in).getStat().getVersion();
-          log.debug("loaded config {} with version {} ", name, zkVersion);
-        }
-        is = new InputSource(in);
-        is.setSystemId(SystemIdResolver.createSystemIdFromResourceName(name));
-      }
 
-      // only enable xinclude, if a SystemId is available
-      if (is.getSystemId() != null) {
-        try {
-          dbf.setXIncludeAware(true);
-          dbf.setNamespaceAware(true);
-        } catch (UnsupportedOperationException e) {
-          log.warn("{} XML parser doesn't support XInclude option", name);
-        }
-      }
-
-      final DocumentBuilder db = dbf.newDocumentBuilder();
-      db.setEntityResolver(new SystemIdResolver(loader));
-      db.setErrorHandler(xmllog);
-      try {
-        doc = db.parse(is);
-        origDoc = doc;
-      } finally {
-        // some XML parsers are broken and don't close the byte stream (but they should according to
-        // spec)
-        IOUtils.closeQuietly(is.getByteStream());
-      }
-      if (substituteProps != null) {
-        DOMUtil.substituteProperties(doc, getSubstituteProperties());
+    try {
+      if (is != null) {
+        doc = SafeXMLParsing.parseConfigXML(log, is, loader);
+      } else {
+        doc = SafeXMLParsing.parseConfigXML(log, loader, name);
```

Comment:
Notice how the parameter order between the method you added and the existing one is kind of jumbled.  Please align; loader before what it is to be loaded..